enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,214 @@
+  - "@dhellmann, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@zshi-redhat, MicroShift "
+  - "@pmtk, MicroShift"
+  - "@mangelajo, MicroShift"
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-01-13
+last-updated: 2023-01-13
+tracking-link:
+  - https://issues.redhat.com/browse/NP-604
+---
+
+# MicroShift Service of Loadbalancer Type Support
+
+## Summary
+
+Customers using loadbalancer type Service don't want to make changes to their
+workloads/deployment for MicroShift as they use other k8s flavors as well. It is
+also promised by us that users can create workload manifests that can work on
+both OpenShift and MicroShift (as long as they do not use APIs unsupported by
+MicroShift).
+
+## Motivation
+
+Normally, the loadbalancer type Service is supported by a Kubernetes cluster
+with external load balancers. In MicroShift, as there is only one node in the
+cluster, we don't really need to have a load balancer. However, from user
+experience perspective, it's important for customers to be able to deploy the
+workload on both OpenShift MicroShift with the same manifests. And there are
+workload types that require a load balancer service (for example, if they are
+not using HTTP/S), so that other features like Ingress or node port service will
+not work. Therefore we need to find a way to support the loadbalancer type
+Service on MicroShift.
+
+### User Stories
+
+As a Microshift user, I want to be able to create a Service manifest with the
+type of `LoadBalancer` and use it on both OCP and MicroShift.  I don't want to
+allocate a new IP address to the Node as the loadbalancer VIP. I want to use the
+NodeIP as the ingress IP of the loadbalancer Service. 
+
+### Goals
+
+* Allow users to deploy the load balancer type Service to MicroShift.
+* The node IP shall be used as the ingress IP of the Service.
+* As we will support pluggable CNI in MicroShift. The implementation shall be
+also compatible with CNIs other than OVN-Kubernetes.
+
+### Non-Goals
+
+* 
+
+## Proposal
+
+To support the loadbalancer type Service, we need to create a new controller
+component that watch the Service objects. It shall be responsible for:
+
+1. work with CNIs to plumb the ingress traffic to the endpoint pods.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+### Workflow Description
+
+**cluster user** is a human user responsible for provisioning Services to a
+MicroShift cluster.
+
+1. The cluster user deploys a Service of type: LoadBalancer 
+2. The loadbalancer service controller watches the creation of the service. It
+will update the Service's `.status.loadBalancer` field, and help the CNIs to 
+plumbing the ingress traffic to the endpoint pods.
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+This new controller is part of the microshift binary. It shall be able to know
+what CNI is used by the cluster, and behave differently based on the CNI type.
```

Comment:
Rather than saying the implementation will have to work with any CNI, let's focus on the CNI we have today and consider expanding to any other implementations to be work to do in the future. When we know we need a different implementation, we could make it so this controller can be disabled and replaced with another one running in a pod, for example, since that would provide complete flexibility.

```suggestion
This new controller is part of the microshift binary. It shall be be limited to the OVN-K
CNI implementation, for now. Updating it to support other CNI implementations will
be left to a future enhancement.
```